contracts/sui/xcall/sources/main.move

```diff
@@ -0,0 +1,547 @@
+    }
+
+
+    fun handle_result(self:&mut Storage,cap:&ConnCap,payload:vector<u8>, ctx: &mut TxContext){
+        let cs_message_result = message_result::decode(&payload);
+        let sequence_no = message_result::sequence_no(&cs_message_result);
+        let code = message_result::response_code(&cs_message_result);
+        let message = message_result::message(&cs_message_result);
+
+        assert!(xcall_state::has_rollback(self, sequence_no), ENoRollback);
+        let rollback = xcall_state::get_rollback(self, sequence_no);
+
+        let sources = rollback_data::sources(&rollback);
+        let to = rollback_data::to(&rollback);
+
+        let source = cap.connection_id();
+
+        let source_valid = is_valid_source(self, to, source, sources);
+
+        assert!(source_valid, EInvalidSource);
+
+        if(vector::length(&sources) > 1){
+            let key = hash::keccak256(&payload);
+            xcall_state::save_pending_responses(self, key, source);
+            let i = 0;
+            if(xcall_state::check_pending_responses(self, key, sources)) return;
+
+            xcall_state::remove_pending_responses(self, key, sources);
+        };
+        event::emit(ResponseMessage { sn: sequence_no, response_code: code });
+
+    if (code == message_result::success()) {
+        xcall_state::remove_rollback(self, sequence_no);
+        if (vector::length(&message) > 0) {
+            let msg = message_request::decode(&message);
+            handle_reply(self,&rollback, &msg, ctx);
+        };
+        xcall_state::set_successful_responses(self, sequence_no);
+    } else {
+        let mut_rollback = xcall_state::get_mut_rollback(self, sequence_no);
+        rollback_data::enable_rollback(mut_rollback);
+        //std::debug::print(&rollback_data::enabled(mut_rollback));
+        event::emit(RollbackMessage{sn:sequence_no, dapp: utils::id_to_hex_string(&rollback_data::from(&rollback)), data: rollback.rollback() });
+    };
+    }
+
+    fun handle_reply(self:&mut Storage, rollback:&RollbackData, reply: &CSMessageRequest, ctx: &mut TxContext){
+        //std::debug::print(&rollback_data::to(rollback));
+        //std::debug::print(&message_request::from_nid(reply) );
+        assert!(rollback_data::to(rollback) == message_request::from_nid(reply), EInvalidReply);
+
+        let req_id = get_next_req_id(self);
+
+        let from = reply.from();
+        let to = reply.to();
+        let data = reply.data();
+        let protocols = reply.protocols();
+        let sn = reply.sn();
+        let msg_type = reply.msg_type();
+
+        let req_id = get_next_req_id(self);
+        let proxy_request = message_request::create(from, to, sn, msg_type, hash::keccak256(&data), protocols);
+        self.add_proxy_request(req_id, proxy_request);
+        event::emit(CallMessage{from,to, sn, req_id,data});
+
+    }
+
+
+    public fun execute_call(self:&mut Storage,cap:&IDCap,request_id:u128,data:vector<u8>,ctx: &mut TxContext):ExecuteTicket{
```

Comment:
Can a dapp here re execute this a claim multiple tickets for the same message?

Reply:
it can get multiple tickets ,  but ticket needs to be resolved in same txn so it will need to call execute result as well which will remove message from queue , unless its persistent msg and result is error which will abort txn anyway.